src/tasks/task_list.hpp

```diff
@@ -27,54 +30,202 @@
 #include "basic_types.hpp"
 #include "task_id.hpp"
 #include "task_types.hpp"
+#include "utils/error_checking.hpp"
 
 namespace parthenon {
 
 enum class TaskListStatus { running, stuck, complete, nothing_to_do };
 
+class TaskList;
+namespace task_list_impl {
+TaskID AddTaskHelper(TaskList *, Task);
+} // namespace task_list_impl
+
+class IterativeTasks {
+ public:
+  IterativeTasks(TaskList *tl, int key) : tl_(tl), key_(key) {
+    max_iterations_ = std::numeric_limits<unsigned int>::max();
+  }
+
+  template <class T, class... Args>
+  TaskID AddTask(TaskID const &dep, T &&func, Args &&... args) {
+    return AddTask_(TaskType::iterative, dep, std::forward<T>(func),
+                    std::forward<Args>(args)...);
+  }
+  // overload to add member functions of class T to task list
+  // NOTE: we must capture the object pointer
+  template <class T, class... Args>
+  TaskID AddTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                 Args &&... args) {
+    return this->AddTask_(TaskType::iterative, dep, [=]() mutable -> TaskStatus {
+      return (obj->*func)(std::forward<Args>(args)...);
+    });
+  }
+
+  template <class T, class... Args>
+  TaskID AddCompletionTask(TaskID const &dep, T &&func, Args &&... args) {
+    return AddTask_(TaskType::completion_criteria, dep, std::forward<T>(func),
+                    std::forward<Args>(args)...);
+  }
+  template <class T, class... Args>
+  TaskID AddCompletionTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                           Args &&... args) {
+    return this->AddTask_(TaskType::completion_criteria, dep,
+                          [=]() mutable -> TaskStatus {
+                            return (obj->*func)(std::forward<Args>(args)...);
+                          });
+  }
+
+  void SetMaxIterations(const unsigned int max) { max_iterations_ = max; }
+  void SetFailWithMaxIterations(const bool flag) { throw_with_max_iters_ = flag; }
+  void SetWarnWithMaxIterations(const bool flag) { warn_with_max_iters_ = flag; }
+  bool ShouldThrowWithMax() const { return throw_with_max_iters_; }
+  bool ShouldWarnWithMax() const { return warn_with_max_iters_; }
+  unsigned int GetMaxIterations() const { return max_iterations_; }
+  unsigned int GetIterationCount() const { return count_; }
+  void IncrementCount() { count_++; }
+
+ private:
+  template <class F, class... Args>
+  TaskID AddTask_(const TaskType &type, TaskID const &dep, F &&func, Args &&... args) {
+    TaskID id(0);
+    id = task_list_impl::AddTaskHelper(
+        tl_, Task(
+                 id, dep,
+                 [=, func = std::forward<F>(func)]() mutable -> TaskStatus {
+                   return func(std::forward<Args>(args)...);
+                 },
+                 type, key_));
```

Comment:
It's just rearranging things, but it seems to me that this bit of nastiness could be hidden by having a task constructor overload that automatically sets the task id to something nonsensical, which we could check for when we validate, to ensure it was overwritten properly.